spotify_tensorflow/dictionary_endpoint.py

```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2018 Spotify AB.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+import logging
+import sys
+import timeit
+from collections import OrderedDict
+
+import six
+import tensorflow as tf
+import numpy as np
+
+from .dataset import Datasets
+
+
+FLAGS = tf.flags.FLAGS
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+
+class DictonaryEndpoint(object):
+    @classmethod
+    def read_dataset(cls, dataset_path, take=sys.maxsize, feature_mapping_fn=None):
+        """
+        Read a TF dataset and load it into a Dictionary of Numpy Arrays.
+
+        :param dataset_path: Path to the TF Records Dataset
+        :param take: Number of records to read
+        :param unpack_multispec: Returns an array of DataFrames, order is the same as the one used
+        when building the MultiFeatureSpec in Featran
+        :param feature_mapping_fn: Override the TF record reading function
+        :return: A Dictonary containing the dataset
+        """
+        return six.next(cls.batch_iterator(dataset_path, take, feature_mapping_fn))
+
+    @classmethod
+    def batch_iterator(cls, dataset_path, batch_size=10000, feature_mapping_fn=None):
+        """
+        Read a TF dataset in batches, each one yielded as a Dictonary.
+
+        :param dataset_path: Path to the TF Records Dataset
+        :param batch_size: Size of each batches
+        :param unpack_multispec: Returns an array of DataFrames, order is the same as the one used
+        when building the MultiFeatureSpec in Featran
+        :param feature_mapping_fn: Override the TF record reading function
+        :return: A Python Generator, yielding batches of data in a Dictonary
+        """
+        training_it, context = Datasets.mk_iter(dataset_path, feature_mapping_fn=feature_mapping_fn)
+        for batch in cls.__FeatureGenerator(training_it, batch_size, context):
+            yield batch
+
+    class __FeatureGenerator(object):
+        def __init__(self, training_it, batch_size, context):
+            self.batch_size = batch_size
+            self.batch_iter = training_it.get_next()
+            self.context = context
+            self.buff = None
+
+        def __iter__(self):
+            logger.info("Starting TF Session...")
```

Comment:
Whenever I see `Starting ... ` comments I like when there is `Closing ...` or `Ending ...` (sometimes with duration) this simplifies debugging. What do you think?

Reply:
Actually I guess, there is `End of dataset`, wonder if we should add duration of the session?